Add limited-reagent dispensers

Partmedia opened the pull request:
## About the PR
This makes chemical, soda, and booze dispensers dispense reagents from a limited stock stored in the machine. Unfortunately, as a result, emagging dispensers for goodies is no longer possible.

This code is backported from Mining Station 14, used with permission, and based on work from @AW-Fulcode.

**Note to mappers:** No action is required. Maps using the existing chemistry/soda/booze dispensers will become the "pre-stocked" version. If you want to change your dispenser to start empty, change your maps to use the empty variants.

## Why / Balance
@EmoGarbage404 wanted this. More discussion here: https://github.com/space-wizards/space-station-14/issues/11345

## Technical details
- Add storage slots to dispensers for the jugs that are stocked
- Populate dispenser inventory from stock stored in machine
- Populate label based on reagent name, or manually-applied label from labeler tool
- Click to transfer from storage unit to container
- Add prototypes for the stock jars of the stuff that should be found in soda and booze dispensers, since some of them were missing

## Media
![image](https://github.com/space-wizards/space-station-14/assets/3229565/5f4b986c-7b11-4585-84bf-3830af6b8629)

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
- `ReagentDispenserComponent`: Remove `EmagPackPrototypeId` field.

:cl: notafet
- tweak: Chemical, booze, and soda dispensers now dispense reagents from stock containers. Remember to refill!
- remove: Chemical, booze, and soda dispensers can no longer be emagged for extra goodies.

Kadeo64 commented:
🍿

EmoGarbage404 commented:
My only suggestion on the ui side would be to either use brackeys or some kind of richtext tag to visually separate the reagent amount from the name.

I think making drinks limited is a bit controversial, but bartenders are usually quite overstocked, so it might be reasonable? 

Make sure the booze/soda dispensers also have empty versions for their machine boards (I couldn't tell if that was implemented or not)

thebadman4662 commented:
Finally antag chems can request uranium for chem to make EMPs and not raise suspicion? Great!

Admiral-Obvious-001 commented:
> Finally antag chems can request uranium for chemlab to make EMPs and not raise suspicion? Great!
> 
> That said, we really need good sources of carbon before its implemented for 15 or so reactions(and thats medical only) that are so bottlenecked by it that not adressing it could start a fire.

Grind some steel.

Partmedia commented:
Yes, I completely forgot about machine construction (and building a machine with magical chemicals). I've fixed that for all three machines.

rbertoche commented:
Nice! Jugs are already introduced for a while now and present in maps, so the game seems ready for it. I was literally wishing for this change on chemistry dispensers in some chat right now!

GreyMario commented:
Regarding booze and soda, I'm a bit yikes on this concept but will ease up on criticism if some method of restocking the machine _a la_ vendors is provided--so that I don't have to buy an entire Booze-o-Mat restock or Bartending Supplies/Softdrinks crate when I only need consumer reagents.

MrGreen006 commented:
I personally don’t see the reason the to remove infinite chems for the bar related dispensers. But refilling is probably not gonna be exactly difficult. The chem dispenser change though I do enjoy, it’s death is ever approaching.

EmoGarbage404 commented:
considering how widespread and global things like decomposition and breaking apart of chemicals into composite parts is going to be for chemistry in the future, I think it's pretty logical to remove all sources of various infinite chemicals. I would be extremely surprised if, in the future, there was no way to extract valuable chemical components from various beverages. Especially since the bar in general is quite well stocked, I don't really have any concerns about the bar going dry and it really is just a safeguard from infinite reagents.

MrGreen006 commented:
> considering how widespread and global things like decomposition and breaking apart of chemicals into composite parts is going to be for chemistry in the future, I think it's pretty logical to remove all sources of various infinite chemicals. I would be extremely surprised if, in the future, there was no way to extract valuable chemical components from various beverages. Especially since the bar in general is quite well stocked, I don't really have any concerns about the bar going dry and it really is just a safeguard from infinite reagents.

Maybe I’m simply overthinking it. I do think this is a neat change that should at the very least be tried. I’m also fairly biased because I haven’t played Lrp in a long time. MRP is mah home.

Aisu9 commented:
> considering how widespread and global things like decomposition and breaking apart of chemicals into composite parts is going to be for chemistry in the future, I think it's pretty logical to remove all sources of various infinite chemicals. I would be extremely surprised if, in the future, there was no way to extract valuable chemical components from various beverages. Especially since the bar in general is quite well stocked, I don't really have any concerns about the bar going dry and it really is just a safeguard from infinite reagents.

aren't these sources be:
botany(plants/fruits/vegetables and medicinal plants)
atmos (for Oxygen,Carbon,Radium,Freezon,Miasma,Plasma)
cargo (every fucking normal chems with the restock box, oxygen and plasma, iron)
Where is bar in the equation? should soda and booze really be on the list of items chemistry can/will have to breakdown for chems?

LankLTE commented:
Given bartender is supposed to be a pretty chill & simple role it seems weird to limit them in the same way you're limiting chem. It's putting pressure on and adding resource management to a role that doesn't really need it imho

GreyMario commented:
> Given bartender is supposed to be a pretty chill & simple role it seems weird to limit them in the same way you're limiting chem. It's putting pressure on and adding resource management to a role that doesn't really need it imho

I can tell there are a good handful of drinks you've never mixed as a bartender, then... 👀

UbaserB commented:
hooray, it is finally here!

thebadman4662 commented:
Where will be chems stored for the purpose of building chemdispenser or it losing the board?

Do built dispensers start empty making Barratry gigapoor in terms of meds and would have to start with one built already? What if someone crowbars it? All gone and reduced to atoms or they are stored in chemboard iself?

MACMAN2003 commented:
this will totally not be a very controversial change :clueless:

EmoGarbage404 commented:
Leaving this comment here before this blows up too much: do not show up here just to spam and doompost. We do not need your special comment about how people are gonna hate this / this is gonna be a shitshow / whatever bullshit you feel like typing.

keep it relevant or i'm just going to lock the PR.

Partmedia commented:
To respond to some of the comments/concerns:

- A button to "dump all chems currently in the beaker back into the dispenser" would reduce the usefulness of electrolysis, centrifuging, and other separation techniques. Besides, there is already a ChemMaster.
- Although there are no "natural" sources for bar drinks, there are plenty of bottles in the vending machines. Individual bottles can be made available for purchase from cargo if we find out that people are indeed running out.
- Chem dispensers currently start "fully stocked".
- The jugs are stored in items slots on the entity, not in the board.

Aisu9 commented:
> To respond to some of the comments/concerns:
> 
>     * A button to "dump all chems currently in the beaker back into the dispenser" would reduce the usefulness of electrolysis, centrifuging, and other separation techniques. Besides, there is already a ChemMaster.
> 
>     * Although there are no "natural" sources for bar drinks, there are plenty of bottles in the vending machines. Individual bottles can be made available for purchase from cargo if we find out that people are indeed running out.
> 
>     * Chem dispensers currently start "fully stocked".
> 
>     * The jugs are stored in items slots on the entity, not in the board.

great, have fun playing Barkeep with 50u bottles in your drinks dispenser. the change for barman serve no value other than adding some realism to the job, it's a spessman social deduction deathmatch game not a survival simulation

thebadman4662 commented:
> > To respond to some of the comments/concerns:
> > ```
> > * A button to "dump all chems currently in the beaker back into the dispenser" would reduce the usefulness of electrolysis, centrifuging, and other separation techniques. Besides, there is already a ChemMaster.
> > 
> > * Although there are no "natural" sources for bar drinks, there are plenty of bottles in the vending machines. Individual bottles can be made available for purchase from cargo if we find out that people are indeed running out.
> > 
> > * Chem dispensers currently start "fully stocked".
> > 
> > * The jugs are stored in items slots on the entity, not in the board.
> > ```
> 
> great, have fun playing Barkeep with 50u bottles in your drinks dispenser. the change for barman serve no value other than adding some realism to the job, it's a spessman social deduction deathmatch game not a survival simulation

I do not see what stops bar dispensers work just like chem dispensers by having jugs/boxes inserted and simple to use.

EmoGarbage404 commented:
> it's a spessman social deduction deathmatch game not a survival simulation

>not a survival simulation

![image](https://github.com/space-wizards/space-station-14/assets/98561806/1bc11fa8-4afd-41ef-a5fe-e226bbaddaad)

idk man according to the [core game design documentation](https://docs.spacestation14.com/en/space-station-14/core-design.html), it seems an awful lot like one.

Admiral-Obvious-001 commented:
You can also add recipes for the common drinks once you run out. It will basically turn into chem but I think fermenting wheat to make beer might be something interesting for the tenders to do once their main stock runs dry, not considering just buying more from cargo.

Aisu9 commented:
> You can also add recipes for the common drinks once you run out. It will basically turn into chem but I think fermenting wheat to make beer might be something interesting for the tenders to do once their main stock runs dry, not considering just buying more from cargo.

this is a wyci issue, and so far no one have been willing to give botany some love, this would also lay more stress on a job designed to be chill. altho thanks god Chemistry don't need botany/atmos/salvage for their job because they can supply all their need at cargo with a request that cost nothing

IamVelcroboy commented:
> You can also add recipes for the common drinks once you run out. It will basically turn into a different flavor of chem more than usual but I think fermenting wheat to make beer might be something interesting for the tenders to do once their main stock runs dry, not considering just buying more from cargo.
> 
> Would give a use for the chefs universal enzyme too for more than just using 5u for cheese.

Time-based reactions for fermenting and a machine for distilling would be awesome!

K-Dynamic commented:
UI is a bit hard on the eyes; maybe shift the container to the right or left of the window

Also, need some differentiation or padding between the reagent buttons

Example:
![image](https://github.com/space-wizards/space-station-14/assets/20566341/1af94f27-c155-4b32-80c6-a6a1efdf2ee7)

Rainbeon commented:
Is the main top-dispenser screen always going to have a scrollbar? Or is it just visible in this particular example where the bottom reagents list is bloated so heavily with many reagents?

Holinka4ever commented:
I understand the thing with chem dispenser
But i am gonna throw my penny into complation about drink/soda dispenser.
In my opinion, they are just drinks, that doesnt do anything impactfull to game while in. Only one drink is doing something, i mean doctor's delight, but its asking chems.
So point is, drinks should gives buffs and debuff, depends which drink it is, cause that state of limit drink, will turn it into water dispenser bar.

Emisse commented:
that is planned for drinks in the future

0x6273 commented:
The reagent and eject buttons should use OpenLeft and OpenRight styles.

deltanedas commented:
if soda were unlimited you could get infinite bicar with the sugar and its elements i think

Holinka4ever commented:
> if soda were unlimited you could get infinite bicar with the sugar and its elements i think


I see.. Knowing that peps mostly munch things out of limit, it will be disaster from prespective. At this point we could clear booze dispence from chem reagents, and try to add food/drink based reagents.



> that is planned for drinks in the future


In same time, its so close, but also far away, a bit of unstable.

VektorZ1 commented:
I would think refilling it via vending machine and jugs would be rather inconvenient

Geekyhobo commented:
Great work only suggestion I have would to be to add some research to the tech tree that allows for upgrades is some format. "Efficient dispensing" or something along those lines (Might be beyond the scope of this PR) I think this is a step to start pushing chem to the next level when it comes to its complexity and systems. People just don't seem to realize that this game is just beginning to touch the level of awesome that it is capable of

Geekyhobo commented:
Also some other comments, I think a more precise dispensing abilities should definitely come with this PR, kind of like custom transfer for beakers and other containers that have the function. Since the chems are finite i think chemist will appreciate the ability to precisely control the chem that they want to dispense. TLDR Keep the quick buttons, and add a box to change to a custom dispensing value.

EmoGarbage404 commented:
We already have droppers and syringes for careful transfer. Being more precise about how you manage chemicals is a good thing. I think it'll make medicine a bit more effortful.

I do agree that chem could totally have research upgrades in the future, but im perfectly fine with the changes in this PR.

SpeltIncorrectyl commented:
Does this mean the chem dispensor can be made easier to get, since it is now just a convenience for mixing chemicals?

Anzarot121 commented:
How should the bartender refill the dispenser? There are special cargo sets for chemists; bartenders do not have one.

Partmedia commented:
- I've adjusted the UI to expand the window so that all buttons are visible (without scrolling).
- I've adjusted the initial labels to use the localized names, so that drinks appear correctly.
- I've added parenthesis around the volume remaining.

Partmedia commented:
Yes, I suppose you could convert the SmartFridge into a "reagent dispenser" that starts empty. As you make medicine, you label those jugs, stick it in the SmartFridge, and it will dispense those drugs. I think you'd be welcome to do that once this is in.

deltanedas commented:
smart fridge literally reskinned chem dispenser :trollface:

0x6273 reviewed (changes requested):
Reagent and eject buttons should use OpenLeft/OpenRight button styles.

0x6273 reviewed (changes requested):
The margin between the reagent button and eject button is too big. It should be same as the amount and clear/eject buttons.

arimah commented:
I'm all in favour of making roundstart dispensers limited in quantity. I must however express an opinion on the chemical grid UI, which is a bit hard to read. As @0x6273, there should definitely be less margin between the reagent button and its associated eject button, to make them visually related. Moreover, the parentheses around the quantities are so visually noisy, and repeated nearly forty times. If I were to design this UI myself, I would make it look more like this:

![image](https://github.com/space-wizards/space-station-14/assets/30327355/ea237b3f-d0c4-4e12-a580-55d41440d442)

The right alignment of the amounts facilitates scanning: one is able more easily to estimate remaining quantity (subconsciously, even, by number of digits), making it trivial to see which reagents are running low; and labels won't jump around as much as chemicals deplete (centered text is generally problematic for text whose length changes a lot).

(Screenshot from an HTML mockup, not in-game.)

lzk228 commented:
just make them use power and have charge

MagnusCrowe commented:
> smart fridge literally reskinned chem dispenser :trollface:

Better than nothing.

rbertoche commented:
Parenthesis did not help, they just take up more space. It was fine without them imo

deltanedas commented:
> just make them use power and have charge

tgats how it used to be, infinite

rbertoche commented:
Just merge it already :P
Is there any reason not to?

EmoGarbage404 commented on Content.Server/Chemistry/Components/ReagentDispenserComponent.cs:
```diff
@@ -7,20 +9,56 @@
+        /// Maximum number of internal storage slots. Dispenser can't store (or dispense) more than
+        /// this many chemicals (without unloading and reloading).
+        /// </summary>
+        [DataField("numStorageSlots")]
```
for most fields here, remove the string from the datafield since it just inherits the variable name in camel case.

EmoGarbage404 commented on Content.Server/Chemistry/Components/ReagentDispenserComponent.cs:
```diff
@@ -7,20 +9,56 @@
+        /// <summary>
+        /// List of storage slots that were created at MapInit.
+        /// </summary>
+        public List<string> StorageSlotIds = new List<string>();
```
it might make sense to make this a datafield since it's only created at mapinit

EmoGarbage404 commented on Content.Server/Chemistry/EntitySystems/ReagentDispenserSystem.cs:
```diff
@@ -80,35 +88,38 @@ private void UpdateUiState(Entity<ReagentDispenserComponent> reagentDispenser)
+                string reagentLabel;
+                if (EntityManager.TryGetComponent<LabelComponent?>(storedContainer, out var label) && label.CurrentLabel != null)
+                    reagentLabel = label.CurrentLabel;
+                else if (EntityManager.TryGetComponent<MetaDataComponent?>(storedContainer, out var metadata))
```
use TryComp instead of EntityManager.TryGetComponent in all the places you have it.

EmoGarbage404 commented on Content.Server/Chemistry/EntitySystems/ReagentDispenserSystem.cs:
```diff
@@ -80,35 +88,38 @@ private void UpdateUiState(Entity<ReagentDispenserComponent> reagentDispenser)
+
+                // Set label from manually-applied label, or metadata if unavailable
+                string reagentLabel;
+                if (EntityManager.TryGetComponent<LabelComponent?>(storedContainer, out var label) && label.CurrentLabel != null)
```
maybe do IsNullOrEmpty for the label check

EmoGarbage404 commented on Content.Server/Chemistry/EntitySystems/ReagentDispenserSystem.cs:
```diff
@@ -80,35 +88,38 @@ private void UpdateUiState(Entity<ReagentDispenserComponent> reagentDispenser)
+                if (EntityManager.TryGetComponent<LabelComponent?>(storedContainer, out var label) && label.CurrentLabel != null)
+                    reagentLabel = label.CurrentLabel;
+                else if (EntityManager.TryGetComponent<MetaDataComponent?>(storedContainer, out var metadata))
+                    reagentLabel = metadata.EntityName;
```
you can just do Name() and skip the trycomp or getting the metadata

EmoGarbage404 commented on Content.Server/Chemistry/EntitySystems/ReagentDispenserSystem.cs:
```diff
@@ -80,35 +88,38 @@ private void UpdateUiState(Entity<ReagentDispenserComponent> reagentDispenser)
+                {
+                    quantity = sol.Volume;
+                }
+                var storedAmount = $"({quantity}u)";
```
put this in locale

EmoGarbage404 reviewed (changes requested):
mostly just conventions
